Tasks/ServiceFabricDeployV1/ServiceFabricSDK/Publish-UpgradedServiceFabricApplication.ps1

```diff
@@ -372,12 +372,80 @@ function Publish-UpgradedServiceFabricApplication
         }
 
         Write-Host (Get-VstsLocString -Key SFSDK_WaitingForUpgrade)
-        $upgradeStatusFetcher = { Get-ServiceFabricApplicationUpgradeAction -ApplicationName $ApplicationName }
+        $unhealthyEvaluationMessageFunction = {
+            param(
+                $UnhealthyEvaluations,
+
+                [string]$Indentation
+            )
+
+            if ($UnhealthyEvaluations -eq $null) {
+                return ""
+            }
+
+            $indentatedErrorString = ""
+            foreach ($UnhealthyEvaluation in $UnhealthyEvaluations) {
+                $indentatedErrorString += $Indentation + $UnhealthyEvaluation.Description
+                # see if indentation needs to be increased. based on the type of evaluation.
+                $indentatedErrorString += & $unhealthyEvaluationMessageFunction $UnhealthyEvaluation, if($UnhealthyEvaluations.kind -eq ($UnhealthyEvaluation.kind + "s")) { return $Indentation } Else { return $Indentation + "`t" }
+            }
+
+            if ($UnhealthyEvaluation.UnhealthyEvent) {
+                $indentatedErrorString += $Indentation + $UnhealthyEvaluation.UnhealthyEvent
+            }
+
+            return $indentatedErrorString
+        }
+
+        $DomainUpgradeStatusFormatterFunction = {
+            param(
+                $UpgradeDomainsStatus
+            )
+
+            if ($UpgradeDomainsStatus -eq $null) {
+                return ""
+            }
+
+            $enumerator = $UpgradeDomainsStatus.GetEnumerator()
+            $upgradeDomainStatusString = ""
+            while ($enumerator.MoveNext() -ne $False) {
+                 $currentElement = $enumerator.Current
+                 $upgradeDomainStatusString += $currentElement.Name + ": " + $currentElement.State + ", "
+            }
+            $upgradeDomainStatusString = $upgradeDomainStatusString.Trim()
+            $upgradeDomainStatusString = $upgradeDomainStatusString.SubString(0, $upgradeDomainStatusString.Length -1 ) #removing last "," from the string.
+            return $upgradeDomainStatusString
+        }
+
+        $upgradeStatusFetcher = {
+            param(
+                $LastUpgradeStatus
+            )
+            $upgradeStatus  = Get-ServiceFabricApplicationUpgrade -ApplicationName $ApplicationName
+            Write-Host "Current Upgrade State:" $upgradeStatus.UpgradeState
+
+            $currentDomainWiseUpgradeStatus = & $DomainUpgradeStatusFormatterFunction $upgradeStatus.UpgradeDomainsStatus
+            $lastDomainWiseUpgradeStatus = if ($LastUpgradeStatus -ne $null) { return & $DomainUpgradeStatusFormatterFunction $LastUpgradeStatus.$UpgradeDomainsStatus } else { return "" }
```

Comment:
We will call omainUpgradeStatusFormatterFunction twice - once for current and once for previous. But that is redundant as it has already been called in previous iteration